ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java

```diff
@@ -749,7 +839,140 @@ private void doContentAddAndReadTest(String digestAlgorithm, ByteBuffer content,
+      doBackPressureTest(digestAlgorithm, content, httpContents, httpContents.size());
+    }
+  }
+
+  /**
+   * Does the backpressure test by ensuring that {@link Channel#read()} isn't called when the number of bytes buffered
+   * is above the {@link NettyRequest#bufferWatermark}. Also ensures that {@link Channel#read()} is called correctly
+   * when the number of buffered bytes falls below the {@link NettyRequest#bufferWatermark}.
+   * @param digestAlgorithm the digest algorithm to use. Can be empty or {@code null} if digest checking is not
+   *                        required.
+   * @param content the complete content.
+   * @param httpContents {@code content} in parts and as {@link HttpContent}. Should contain all the data in
+   * {@code content}.
+   * @param numChunksToAddBeforeRead the number of {@link HttpContent} to add before making the
+   * {@link NettyRequest#readInto(AsyncWritableChannel, Callback)} call.
+   * @throws Exception
+   */
+  private void doBackPressureTest(String digestAlgorithm, ByteBuffer content, List<HttpContent> httpContents,
+      int numChunksToAddBeforeRead)
+      throws Exception {
+    if (numChunksToAddBeforeRead < 0 || numChunksToAddBeforeRead > httpContents.size()) {
+      throw new IllegalArgumentException("Illegal value of numChunksToAddBeforeRead");
+    }
+
+    MockChannel channel = new MockChannel();
+    int expectedMaxMessagesPerRead = channel.config().getMaxMessagesPerRead();
+    final NettyRequest nettyRequest = createNettyRequest(HttpMethod.POST, "/", null, channel);
+    byte[] wholeDigest = null;
+    if (digestAlgorithm != null && !digestAlgorithm.isEmpty()) {
+      MessageDigest digest = MessageDigest.getInstance(digestAlgorithm);
+      digest.update(content);
+      wholeDigest = digest.digest();
+      content.rewind();
+      nettyRequest.setDigestAlgorithm(digestAlgorithm);
+    }
+
+    final AtomicInteger queuedReads = new AtomicInteger(0);
+    ByteBufferAsyncWritableChannel writeChannel = new ByteBufferAsyncWritableChannel();
+    ReadIntoCallback callback = new ReadIntoCallback();
+
+    channel.setChannelReadCallback(new MockChannel.ChannelReadCallback() {
+      @Override
+      public void onRead() {
+        queuedReads.incrementAndGet();
+      }
+    });
+
+    int addedCount = 0;
+    Future<Long> future = null;
+    boolean suspended = false;
+    int bytesToVerify = 0;
+    while (addedCount < httpContents.size()) {
+      if (suspended) {
+        assertEquals("There should have been no reads queued when over buffer watermark", 0, queuedReads.get());
+        if (future == null) {
+          future = nettyRequest.readInto(writeChannel, callback);
+        }
+        int chunksRead = readAndVerify(bytesToVerify, writeChannel, content);
+        assertEquals("Number of reads triggered is not as expected", chunksRead, queuedReads.get());
+        // collapse many reads into one
+        queuedReads.set(1);
```

Comment:
This collapses many reads on the read channel into one read. This read is for the read/socket channel and not for the write channel (which like you said has nothing queued)